### PR TITLE
kubebuilder v4 + dependencies upgrade

### LINKED ISSUE #908: Update Kubebuilder go plugin
Operator currently uses go.kubebuilder.io/v2 plugin which is already deprecated, would be great to update to v3 or v4.
>[Deprecation Notice] This version is deprecated and is no longer scaffolded by default since `28 Apr 2021`.The `go/v2` plugin cannot scaffold projects in which CRDs and/or Webhooks have a `v1` API version.Be aware that v1beta1 API for CRDs and Webhooks was deprecated on Kubernetes 1.16 and areremoved as of the Kubernetes 1.22 release. Therefore, since this plugin cannot produce projects thatwork on Kubernetes versions >= 1.22, it is recommended to upgrade your project to the latest versions available.

Two ways to implement:
1. create a new v3 project and copy over the API and the reconciliation code, see https://book.kubebuilder.io/migration/legacy/migration_guide_v2tov3
2. updating all the files manually, see https://book.kubebuilder.io/migration/legacy/manually_migration_guide_v2_v3#migration-from-v2-to-v3-by-updating-the-files-manually

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
- kubebuilder v2 -> v4
- upgraded dependencies

Fixes #908

### CODE REVIEW

File: internal/manager/manager.go [hunk 1]
```diff
@@ -53,14 +53,14 @@ var (
 	})
 	scheme               = runtime.NewScheme()
 	setupLog             = ctrl.Log.WithName("setup")
-	enableLeaderElection = managerFlags.Bool("enable-leader-election", false, "Enable leader election for controller manager. "+
+	enableLeaderElection = managerFlags.Bool("leader-elect", false, "Enable leader election for controller manager. "+
```
Comment (Haleygo) [inline internal/manager/manager.go]: Why change those flags? it brings breaking changes.

Comment (AndrewChubatiuk) [reply internal/manager/manager.go]: it's a part of v2-> v3 upgrade
https://kubebuilder.io/migration/legacy/manually_migration_guide_v2_v3#rename-the-manager-flags

Comment (Haleygo) [reply internal/manager/manager.go]: Could you mention all the breaking changes in the changelog, and provide update notes?

Comment (f41gh7) [reply internal/manager/manager.go]: I'm not happy with this change. `enableLeaderElection` is commonly set flag and it ll break a lot of operator deployments. 

 I'd like to postpone it and address it to operator v1 release.

Comment (AndrewChubatiuk) [reply internal/manager/manager.go]: @f41gh7 what is a problem with it? people still rely on helm chart, where this flag can be changed without any issues for them. also to support old and new operator versions in helm chart i could add there a condition, which will sets flag names depending on a version

Comment (f41gh7) [reply internal/manager/manager.go]: My concern is:
- version of chart and operator could managed separately. And you cannot simply rollback chart to previous image version
- it's not reliable to detect operator version by image tag
- not everyone uses helm chart for operator deployment

 Maybe it's fine to mention it at changelog. Until v1 version operator doesn't guarantee backward compatibility.

Comment (f41gh7) [reply internal/manager/manager.go]: Lets change it, without it migration wouldn't be completed.

**Review by f41gh7 — DISMISSED** [review]
LGTM

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,18 @@ aliases:
...
+### Breaking changes
+
+**Update note 1: the `--metrics-addr` command-line flag at `operator` was deprecated. Use `--metrics-bind-address` instead.**
+**Update note 2: the `--enable-leader-election` command-line flag at `operator` was deprecated. Use `--leader-elect` instead.**
+
+### Features
+
+- [operator](./README.md): kubebuilder v2 -> v4 upgrade
+- [operator](./README.md): added support of alertmanagerconfigs.monitoring.coreos.com/v1beta1
+- [operator](./README.md): upgraded customresourcedefinitions.apiextensions.k8s.io/v1beta1 to customresourcedefinitions.apiextensions.k8s.io/v1
```
Comment (Haleygo) [inline docs/CHANGELOG.md]: We don't use `customresourcedefinitions.apiextensions.k8s.io/v1beta1` as it's already deprecated, this [patches dir](https://github.com/VictoriaMetrics/operator/tree/master/config/crd/patches) is not used before as I can see.

Comment (AndrewChubatiuk) [reply docs/CHANGELOG.md]: added [PR](https://github.com/VictoriaMetrics/operator/pull/985) which removes patches directory

Comment (AndrewChubatiuk) [reply docs/CHANGELOG.md]: removed this from changelog

File: internal/controller/vmprometheusconverter_controller.go [hunk 3]
```diff
@@ -151,46 +161,86 @@ func NewConverterController(ctx context.Context, baseClient *kubernetes.Clientse
...
 	}); err != nil {
 		return nil, fmt.Errorf("cannot add alertmanager_config handler: %w", err)
 	}
+	c.amConfigInfs = append(c.amConfigInfs, informer{
+		informer: amConfigInf,
+		version:  promv1alpha1.SchemeGroupVersion.String(),
+		kind:     promv1alpha1.AlertmanagerConfigKind,
+	})
+
+	// AlertmanagerConfig/v1beta1
```
Comment (Haleygo) [inline internal/controller/vmprometheusconverter_controller.go]: Can we check if cluster has installed prometheus v1beta1 CRD first, and skip watching if there is not?
Now running with env that doesn't install v1beta1.AlertmanagerConfig gives errors
```
{"level":"info","ts":"2024-06-20T17:22:14+08:00","msg":"pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:232: failed to list *v1beta1.AlertmanagerConfig: cannot list alertmanager_configs: cannot list objects at cluster scope: no kind is registered for the type v1beta1.AlertmanagerConfigList in scheme \"pkg/runtime/scheme.go:100\""}
{"level":"error","ts":"2024-06-20T17:22:14+08:00","msg":"pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:232: Failed to watch *v1beta1.AlertmanagerConfig: failed to list *v1beta1.AlertmanagerConfig: cannot list alertmanager_configs: cannot list objects at cluster scope: no kind is registered for the type v1beta1.AlertmanagerConfigList in scheme \"pkg/runtime/scheme.go:100\"","stacktrace":"k8s.io/klog/v2.(*loggingT).output\n\t/Users/haley/go/pkg/mod/k8s.io/klog/v2@v2.120.1/klog.go:912\nk8s.io/klog/v2.(*loggingT).printWithInfos\n\t/Users/haley/go/pkg/mod/k8s.io/klog/v2@v2.120.1/klog.go:740\nk8s.io/klog/v2.(*loggingT).printDepth\n\t/Users/haley/go/pkg/mod/k8s.io/klog/v2@v2.120.1/klog.go:722\nk8s.io/klog/v2.ErrorDepth\n\t/Users/haley/go/pkg/mod/k8s.io/klog/v2@v2.120.1/klog.go:1591\nk8s.io/apimachinery/pkg/util/runtime.logError\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/runtime/runtime.go:115\nk8s.io/apimachinery/pkg/util/runtime.HandleError\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/runtime/runtime.go:109\nk8s.io/client-go/tools/cache.DefaultWatchErrorHandler\n\t/Users/haley/go/pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:150\nk8s.io/client-go/tools/cache.(*Reflector).Run.func1\n\t/Users/haley/go/pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:299\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/wait/backoff.go:226\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/wait/backoff.go:227\nk8s.io/client-go/tools/cache.(*Reflector).Run\n\t/Users/haley/go/pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:297\nk8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/wait/wait.go:55\nk8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/wait/wait.go:72"}
{"level":"info","ts":"2024-06-20T17:23:03+08:00","msg":"pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:232: failed to list *v1beta1.AlertmanagerConfig: cannot list alertmanager_configs: cannot list objects at cluster scope: no kind is registered for the type v1beta1.AlertmanagerConfigList in scheme \"pkg/runtime/scheme.go:100\""}
{"level":"error","ts":"2024-06-20T17:23:03+08:00","msg":"pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:232: Failed to watch *v1beta1.AlertmanagerConfig: failed to list *v1beta1.AlertmanagerConfig: cannot list alertmanager_configs: cannot list objects at cluster scope: no kind is registered for the type v1beta1.AlertmanagerConfigList in scheme \"pkg/runtime/scheme.go:100\"","stacktrace":"k8s.io/klog/v2.(*loggingT).output\n\t/Users/haley/go/pkg/mod/k8s.io/klog/v2@v2.120.1/klog.go:912\nk8s.io/klog/v2.(*loggingT).printWithInfos\n\t/Users/haley/go/pkg/mod/k8s.io/klog/v2@v2.120.1/klog.go:740\nk8s.io/klog/v2.(*loggingT).printDepth\n\t/Users/haley/go/pkg/mod/k8s.io/klog/v2@v2.120.1/klog.go:722\nk8s.io/klog/v2.ErrorDepth\n\t/Users/haley/go/pkg/mod/k8s.io/klog/v2@v2.120.1/klog.go:1591\nk8s.io/apimachinery/pkg/util/runtime.logError\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/runtime/runtime.go:115\nk8s.io/apimachinery/pkg/util/runtime.HandleError\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/runtime/runtime.go:109\nk8s.io/client-go/tools/cache.DefaultWatchErrorHandler\n\t/Users/haley/go/pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:150\nk8s.io/client-go/tools/cache.(*Reflector).Run.func1\n\t/Users/haley/go/pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:299\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/wait/backoff.go:226\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/wait/backoff.go:227\nk8s.io/client-go/tools/cache.(*Reflector).Run\n\t/Users/haley/go/pkg/mod/k8s.io/client-go@v0.30.2/tools/cache/reflector.go:297\nk8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/wait/wait.go:55\nk8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1\n\t/Users/haley/go/pkg/mod/k8s.io/apimachinery@v0.30.2/pkg/util/wait/wait.go:72"}
```
And looks like it should be added as a feature in changelog as well.

Comment (AndrewChubatiuk) [reply internal/controller/vmprometheusconverter_controller.go]: removed v1beta1.AlertmanagerConfig informer, as it's CRDs are not available yet and just updated code to stop using it

Comment (Haleygo) [reply internal/controller/vmprometheusconverter_controller.go]: Okay, so prometheus created v1beta1 AlertmanagerConfig but still [not migrate to it](https://github.com/prometheus-community/helm-charts/pull/3938) in prometheus-operator or their chart.
Could you also remove other codes which using `promv1beta1.AlertmanagerConfig`
https://github.com/VictoriaMetrics/operator/blob/232d977c68aff75efa7e7b53d3273fd4896c19b1/internal/controller/vmprometheusconverter_controller.go#L487-L488

File: docs/CHANGELOG.md [hunk 4]
```diff
@@ -16,6 +16,16 @@ aliases:
 
 - [operator](#./README.md): fix VM CRs' `xxNamespaceSelector` and `xxSelector` options, previously they are inverted. See this [issue](https://github.com/VictoriaMetrics/operator/issues/980) for details.
 
+### Breaking changes
```
Comment (Haleygo) [inline docs/CHANGELOG.md:17]: Could you move this to the top of the changelog?

Comment (AndrewChubatiuk) [reply docs/CHANGELOG.md:17]: done

File: internal/controller/vmprometheusconverter_controller.go (hunk 3 above)
Comment (AndrewChubatiuk) [reply internal/controller/vmprometheusconverter_controller.go]: done

**Review by Haleygo — APPROVED** [review]
LGTM

**Comment by Haleygo** [comment]
Thank you!

**Pull request merged**